### ISSUE TITLE
plugin seems to be going to sleep/hanging after a while on M1 mac 

### ISSUE THREAD

**Issue by dan1916** [ebaauw/homebridge-sc#9]
Hello,

First of all, thank you so much for developing this fantastic plugin. It is game changer for me since I have more than 10 soma shades and the functionality provided by this plugin is far more than soma connect's native homekit one. I find the light sensor values specially useful. 

I have been running into this issue where all the shades stops responding in the home app or in automations after a while. For example, if I don't touch homebridge (like login into it, view logs, reboot it etc), after a day, the blinds exposed by this plugin seems to be going into some sort of sleep state. In the home app, it keeps on spinning the wheel when I try to close or open them. It does not show the red "no response", it just keeps spinning the gray wheel. Sunset/Sunrise automations does not execute. It is only when I reboot the child bridge of this plugin or reboot the whole homebridge that it starts working. And surprisingly it seems to be 'remembering' the last command that was sent, so it just replays that. i..e. if it was stuck on opening of blinds in the morning, if I reboot the homebridge in the afternoon, it will open the blinds soon after.

This is only happening with this plugin. I have many other plugins as well as native homekit accessories and all of them are working fine. I never have to reboot homebridge for anything. I am running it on a M1 Mac Mini with MacOS Ventura 13.1. The Mac mini has a fixed static IP. The soma connect, i.e. my RaspberryPi 3 stays on during these times, and fully functional also with a static IP. I can open up various API end points successfully in a browser on a separate machine. It is just this plugin which seems to be get stuck/waiting/sleeping on something.

I had identical issue back in May 2022 and I had to uninstall the plugin back then and jump on to the soma Connect's homekit bridge. But that is very limiting. I want to use this plugin in all seriousness. It provides so much information which is missing in soma's implementation. Is this something that can be debugged or fixed? I can help in getting any further information that you need.

Thanks in advance!

**Comment by ebaauw** [comment]
Are you running other Homebridge instances on the Mac?  It sounds like the Mac might have gone to sleep.

Apple’s Home app acts a little weird for _Window Covering_ devices, especially while the _Current Position_ doesn’t reflect the _Target Position_.  You might want to refresh or even force-quit and re-open Home.  Can you still control the blinds from another HomeKit app (like Eve)?
Does the spinning occur on one Apple device (iPhone, Mac, …) or on multiple?  Does it stop when you power cycle your home hub (AppleTV, HomePod).

To analyse whether this is caused by the plugin, please run Homebridge in DEBUG mode, make sure _Log Level_ on the SOMA Connect accessory is set to 2 (in Eve or another decent HomeKit app), and capture and attach the Homebridge log file.

**Comment by dan1916** [comment]
This is the only homebridge instance I am running on the Mac. The Mac stays awake all the time, I have set the power settings to maximum performance and never sleep.

The issue is not only in the Home app I would say, it is when the automations execute. Once the issue happens, automations are not able to close/open blinds on their time. Using a third party Home app like Eve, the blinds show the spinning bar when asked to close or open by clicking on it or via an automation. The same behavior of spinning happens on my iPhone and the Mac's Home app, too.

I haven't tried setting the log level yet, I can try that and report back. I suspect the Raspberry pi 3's soma connect image is not able to handle simultaneous connections to more than 10 blinds when homekit asks it.

**Comment by ebaauw** [comment]
What automations do you use?  On-device through the SOMA app?  HomeKit automations?  Or an external home automation system?

Spinning bar sounds like HomeKit is waiting for a response from Homebridge.  It should revert to _No Response_ if the connection is lost or Homebridge has crashed.  I really need to see the Homebridge log file to ascertain whether it's an issue with Homebridge SC, or with your HomeKit setup.  Did you try and power-cycle the home hub when the spinning is happening?

From what I can tell, the service running on the SOMA connect is a NodeJS programme, bundled with Noble (for BLE communication) and HAP-NodeJS (for the native HomeKit feature).  In my experience, using concurrent sessions to multiple BLE devices is very challenging for Noble.  I ended up serialising everything in Homebridge SOMA, having only one connection at a time, but even that wouldn't result in the stability I wanted.  The service on the SOMA Connect crashes occasionally, but is restarted automatically.  Homebridge SC should reconnect to the restarted service.

Note that a SOMA devices only allows one connection at a time.  The Connect cannot control the device, while you're controlling it through the SOMA app.  Also the SOMA device stops advertising when it has an active connection.  Position and battery level are included in the BLE advertisements, so the Connect refreshes these without connecting to the device, but a connection is needed for controlling the device and reading the light level.

**Comment by dan1916** [comment]
I am using HomeKit automations.

I did try to reboot homehub while the spinning happens, it shows No response for a minute, then everything else updates their status but the blinds come back to spinning.

I really wish Soma had either developed a better software for their bridge or made it open source so other can contribute and improve it. Controlling however number of blinds from my Soma iOS app always works 100%. I have created groups and one group has 8 blinds in it. iOS app takes about 3-4 seconds to connect them all and then with one click I can raise or lower them all, at the same time. iOS app never fails to connect. 

I tried running Hombridge SC on a fresh Ubuntu box with fresh Homebridge. I crashes the HTTP server on my Soma Connect rather quickly. Within a minute I start seeing lines like connection refused, and then if I try to open /list_devices from any device, it does not load. I have no idea what it does to the HTTP server. Is there a way to see logs of any sort on the Soma Connect?

Also is there a way to expose just the light sensors from this plugin? In that case I can use soma connect's homekit for control and this plugin to get light levels and execute light based automations. Why wouldn't soma connect simply give the light levels as well?

**Comment by ebaauw** [comment]
>  I crashes the HTTP server on my Soma Connect rather quickly.

That’s what I said earlier: the service crashes, but should be restarted automatically. 

> Is there a way to see logs of any sort on the Soma Connect?

You can login to the box, see https://support.somasmarthome.com/hc/en-us/articles/360014397380-SOMA-Connect-Raspberry-Pi-username-and-password, but I haven’t found any extensive logging.  You can enable SSH simply by creating a `ssh` file on the `/boot` partition of the image.

> In that case I can use soma connect's homekit for control

Have you tried controlling multiple HomeKit accessories exposed natively by the SOMA Connect?  E.g. in a HomeKit scene?  I wouldn’t be surprised if you run into similar issues.

> Also is there a way to expose just the light sensors from this plugin?

No.  Homebridge SC doesn’t expose a separate light sensor service, because the reported light value cannot be normalised to lux.  It just exposes a current ambient light level characteristic with the raw value on the window covering service.

> this plugin to get light levels and execute light based automations

What are you automating?  If it’s just the blinds themselves, you can use on-device automations.  Then you can use HomeKit to trigger other accessories on the blinds position.

**Comment by dan1916** [comment]
A bit of update on this one. I did two things. First, I did fresh install on Ubuntu machine I had. That worked most times, and never showed a spinning while (unlike mac), but it ran into connection refused errors pretty soon with the HTTP server on soma connect crashing for whatever reason.

Then I unpaired the homekit from soma connect and just used the homekit integration provided by this plugin.

Finally experimented with connection the pi with ethernet only and disabled the wifi radio. I think that seem to have helped a lot. The connection is pretty stable at the moment. It is able to control 10+ blinds very fast every single time since last 3 days. I will let it run for a few more days to check the stability. Feel free to close this issue for now. Thank you for your detailed answer and an awesome plugin!
